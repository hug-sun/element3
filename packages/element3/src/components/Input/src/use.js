import { ref, computed, watch, reactive, toRefs, inject, onMounted } from 'vue'
import calcTextareaHeight from './calcTextareaHeight'

export const useClass = (inputSize, inputExceed, props, attrs, slots) => {
  return computed(() => {
    const classes = [
      inputSize.value ? 'el-input--' + inputSize.value : '',
      props.type === 'textarea' ? 'el-textarea' : 'el-input'
    ]

    if (attrs.disabled) {
      classes.push('is-disabled')
    }

    if (inputExceed.value) {
      classes.push('is-exceed')
    }

    if (slots.prepend || slots.append) {
      classes.push('el-input-group')
    }

    if (slots.append) {
      classes.push('el-input-group--append')
    }

    if (slots.prepend) {
      classes.push('el-input-group--prepend')
    }
    if (slots.prefix || props.prefixIcon) {
      classes.push('el-input--prefix')
    }

    if (
      slots.suffix ||
      props.suffixIcon ||
      slots.suffixIcon ||
      attrs.clearable ||
      attrs.showPassword
    ) {
      classes.push('el-input--suffix')
    }

    return classes
  })
}

export const useInput = (props, cxt, textarea) => {
  const input = ref(null)

  const elFormItem = inject('elFormItem', {})

  const elFormChange = inject('elForm.change', () => {})

  const {
    modelValue,
    size,
    suffixIcon,
    clearable,
    showPassword,
    showWordLimit
  } = toRefs(props)

  const nativeInputValue = computed(() => {
    return modelValue.value === null || modelValue.value === undefined
      ? ''
      : String(modelValue.value)
  })

  const textLength = computed(() => {
    if (typeof modelValue.value === 'number') {
      return String(modelValue.value).length >= Number(cxt.attrs.maxlength)
        ? Number(cxt.attrs.maxlength)
        : String(modelValue.value).length
    }

    return modelValue.value.length >= Number(cxt.attrs.maxlength)
      ? Number(cxt.attrs.maxlength)
      : modelValue.value.length
  })

  const elFormItemSize = computed(() => {
    return elFormItem.elFormItemSize || ''
  })

  const inputSize = computed(() => {
    return size?.value || elFormItemSize.value
  })

  const inputExceed = computed(() => {
    return props.modelValue?.length >= Number(cxt.attrs.maxlength)
      ? true
      : false
  })

  const getInput = () =>
    computed({
      get: () => (input.value ? input.value : textarea.value),
      set: (value) => {
        if (input?.value) {
          input.value.value = value
        }

        if (textarea?.value) {
          textarea.value.value = value
        }
      }
    })

  const clearValue = () => {
    cxt.emit('update:modelValue', '')
  }

  const setNativeInputValue = () => {
    const input = getInput()
    if (!input) return
    input.value = nativeInputValue.value
  }

  const getSuffixVisible = computed(() => {
    return (
      cxt.slots.suffix ||
      suffixIcon?.value ||
      clearable?.value ||
      showPassword?.value ||
      showWordLimit?.value
    )
  })

  onMounted(() => {
    setNativeInputValue(nativeInputValue.value)
  })

  watch(
    () => props.modelValue,
    () => {
      setNativeInputValue()
      if (props.validateEvent) {
        elFormChange()
      }
    }
  )

  return {
    input,
    getInput,
    nativeInputValue,
    textLength,
    clearValue,
    inputSize,
    getSuffixVisible,
    inputExceed
  }
}

export const useInputEvent = (emit) => {
  const handleInput = (event) => {
    emit('update:modelValue', event.target.value)
    emit('input', event.target.value)
  }

  const handleFocus = (event) => {
    emit('focus', event)
  }

  const handleBlur = (event) => emit('blur', event)

  const handleClear = () => {
    emit('update:modelValue', '')
    emit('clear')
  }

  const onChange = (event) => {
    emit('change', event.target.value)
  }

  return {
    handleInput,
    handleFocus,
    handleBlur,
    handleClear,
    onChange
  }
}

export const useInputMethod = (element) => {
  const focus = () => {
    element.value.focus()
  }

  const blur = () => {
    element.value.blur()
  }

  const select = () => {
    element.value.select()
  }

  return {
    focus,
    blur,
    select
  }
}

export const useTextarea = (props) => {
  const textarea = ref(null)

  const { autosize, type, resize } = toRefs(props)

  const state = reactive({
    textareaCalcStyle: {}
  })

  const textareaStyle = computed(() => {
    return Object.assign({}, state.textareaCalcStyle, { resize: resize?.value })
  })

  watch(
    () => props.modelValue,
    () => resizeTextarea()
  )

  const resizeTextarea = () => {
    if (type.value !== 'textarea') return
    if (!autosize.value) {
      state.textareaCalcStyle = {
        minHeight: calcTextareaHeight(textarea.value).minHeight
      }

      return
    }
    const minRows = autosize.value.minRows
    const maxRows = autosize.value.maxRows
    state.textareaCalcStyle = calcTextareaHeight(
      textarea.value,
      minRows,
      maxRows
    )
  }

  return { textarea, textareaStyle, resizeTextarea }
}
