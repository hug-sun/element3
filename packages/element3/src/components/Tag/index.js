import ElTag from './src/Tag.vue'

/* istanbul ignore next */
ElTag.install = function (app) {
  app.component(ElTag.name, ElTag)
}

export { ElTag }
