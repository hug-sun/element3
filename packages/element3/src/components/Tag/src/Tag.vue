<script lang="jsx">
import { useGlobalOptions } from '../../../composables/globalConfig.js'
import { ref, computed, toRefs, Transition } from 'vue'

const useShow = (emit) => {
  const show = ref(true)

  const handleClose = (e) => {
    e.stopPropagation()
    show.value = false
    emit('close', e)
  }

  return {
    show,
    handleClose
  }
}
export default {
  name: 'ElTag',
  props: {
    type: {
      type: String,
      default: '',
      validator(v) {
        return ['success', 'info', 'warning', 'danger', ''].includes(v)
      }
    },
    closable: Boolean,
    disableTransitions: Boolean,
    hit: Boolean,
    color: String,
    size: {
      type: String,
      validator(v) {
        return ['medium', 'small', 'mini', ''].includes(v)
      }
    },
    effect: {
      type: String,
      default: 'light',
      validator(v) {
        return ['dark', 'light', 'plain'].includes(v)
      }
    }
  },
  emits: ['close', 'click'],
  setup(props, { emit }) {
    const $ELEMENT = useGlobalOptions()

    const { type, hit, size, effect } = toRefs(props)

    const tagSize = computed(() => {
      return size?.value || $ELEMENT.size
    })
    const classes = computed(() => [
      'el-tag',
      type.value ? `el-tag--${type.value}` : '',
      tagSize.value ? `el-tag--${tagSize.value}` : '',
      `el-tag--${effect.value}`,
      hit.value && 'is-hit'
    ])

    const { show, handleClose } = useShow(emit)

    return {
      show,
      classes,
      handleClose
    }
  },
  render() {
    const tagEl = this.show ? (
      <span
        class={this.classes}
        style={{ backgroundColor: this.color }}
        onClick={(e) => {
          this.$emit('click', e)
        }}
      >
        {this.$slots.default?.()}
        {this.closable && (
          <i class="el-tag__close el-icon-close" onClick={this.handleClose} />
        )}
      </span>
    ) : (
      ''
    )
    return this.disableTransitions ? (
      tagEl
    ) : (
      <Transition appear name="el-zoom-in-center">
        {tagEl}
      </Transition>
    )
  }
}
</script>
