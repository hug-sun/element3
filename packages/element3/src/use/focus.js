import { getCurrentInstance, isRef } from 'vue'

function useFocus(ref) {
  return function focus() {
    if (isRef(ref)) {
      ref.value.focus()
    } else {
      const { proxy } = getCurrentInstance()
      proxy.$refs[ref].focus()
    }
  }
}

export default useFocus
