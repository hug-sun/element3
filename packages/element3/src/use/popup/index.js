import PopupComponent from './PopupComponent'
// import './style.scss'
export { PopupComponent }
