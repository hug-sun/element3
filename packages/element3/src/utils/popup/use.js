import { computed, onBeforeMount, onBeforeUnmount, onUnmounted } from 'vue'
import { addClass, removeClass, hasClass } from '../../utils/dom'

let zIndex = 2000

let useBodyScrollCounter = 0

function useZindex() {
  return computed(() => zIndex++)
}

function useBodyScroll({ lockScroll }) {
  const body = document.body

  useBodyScrollCounter++
  onBeforeMount(() => {
    if (lockScroll && !hasClass(body, 'el-popup-parent--hidden')) {
      addClass(body, 'el-popup-parent--hidden')
    }
  })

  onUnmounted(() => {
    useBodyScrollCounter--
    if (lockScroll || useBodyScrollCounter <= 1) {
      removeClass(document.body, 'el-popup-parent--hidden')
    }
  })
}

export { useZindex, useBodyScroll }
