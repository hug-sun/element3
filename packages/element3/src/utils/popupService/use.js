import { computed, onBeforeMount, onUnmounted, ref } from 'vue'
import { addClass, removeClass, hasClass } from '../dom'

let zIndex = 2000

let useBodyScrollCounter = 0

function useZindex() {
  return zIndex++
}

export const useStyle = () => {
  const style = ref({})

  function setStyle(options) {
    style.value = Object.assign({}, style, options)
  }
  return {
    style,
    setStyle
  }
}

function useBodyScroll({ lockScroll }) {
  const body = document.body

  onBeforeMount(() => {
    if (lockScroll) {
      useBodyScrollCounter++
      if (!hasClass(body, 'el-popup-parent--hidden')) {
        addClass(body, 'el-popup-parent--hidden')
      }
    }
  })

  onUnmounted(() => {
    if (lockScroll) {
      useBodyScrollCounter--
      if (useBodyScrollCounter <= 0) {
        removeClass(document.body, 'el-popup-parent--hidden')
      }
    }
  })
}

export { useZindex, useBodyScroll }
