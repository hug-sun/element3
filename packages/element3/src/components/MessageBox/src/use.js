import { toRefs, nextTick, unref, onMounted, onUnmounted, computed } from 'vue'
import { usePopup } from '../../../composables/popup'
import { isFunction } from '@vue/shared'
export function useHandleList(state, instance, validate) {
  const { close } = usePopup(state)
  const {
    visible,
    callback,
    category,
    inputType,
    distinguishCancelAndClose,
    closeOnClickModal,
    action,
    beforeClose
  } = toRefs(state)
  const closeHandle = () => {
    visible.value = false
    close()
    nextTick(() => {
      unref(callback)(state.action, instance.proxy)
    })
  }
  const handleAction = (_action) => {
    if (unref(category) === 'prompt' && _action === 'confirm' && !validate()) {
      return
    }
    action.value = _action
    if (isFunction(beforeClose.value)) {
      beforeClose.value(_action, instance.proxy, closeHandle)
    } else {
      closeHandle()
    }
  }

  const handleInputEnter = () => {
    if (unref(inputType) !== 'textarea') {
      return handleAction('confirm')
    }
  }

  const handleWrapperClick = () => {
    if (unref(closeOnClickModal)) {
      handleAction(unref(distinguishCancelAndClose) ? 'close' : 'cancel')
    }
  }

  return {
    closeHandle,
    handleAction,
    handleWrapperClick,
    handleInputEnter
  }
}

export function watchElement(state, handleAction, closeHandle) {
  const {
    closeOnHashChange,
    closeOnPressEscape,
    distinguishCancelAndClose
  } = toRefs(state)
  const { open } = usePopup(state)
  const handleKeyup = (element = {}) => {
    if (element.code !== 'Escape') return
    if (unref(closeOnPressEscape)) {
      handleAction(unref(distinguishCancelAndClose) ? 'close' : 'cancel')
    }
  }
  onMounted(() => {
    if (unref(closeOnHashChange)) {
      window.addEventListener('hashchange', closeHandle)
    }
    window.addEventListener('keyup', handleKeyup)
    nextTick(() => {
      open()
    })
  })
  onUnmounted(() => {
    if (unref(closeOnHashChange)) {
      window.removeEventListener('hashchange', closeHandle)
    }
    window.removeEventListener('keyup', handleKeyup)
  })
}

export function classIcon(iconClass, type) {
  return computed(() => {
    return unref(iconClass) || (unref(type) ? `el-icon-${unref(type)}` : '')
  })
}
