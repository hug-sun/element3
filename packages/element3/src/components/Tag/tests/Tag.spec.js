import Tag from '../src/Tag.vue'
import { mount } from '@vue/test-utils'
import { ref, nextTick } from 'vue'
import { setupGlobalOptions } from '../../../use/globalConfig'

describe('Tag.vue', () => {
  it('normal render', () => {
    const wrapper = mount(Tag)
    const result = wrapper.findComponent(Tag).exists()

    expect(result).toBeTruthy()
  })
  it('content', () => {
    const wrapper = mount(Tag, {
      slots: {
        default: 'tag'
      }
    })
    const result = wrapper.findComponent(Tag).text()

    expect(result).toBe('tag')
  })
  describe('props', () => {
    it('type', () => {
      const type = 'success'
      const wrapper = mount(Tag, {
        props: {
          type
        }
      })
      const result = wrapper
        .findComponent(Tag)
        .find(`.el-tag--${type}`)
        .exists()

      expect(result).toBeTruthy()
    })
    it('closable', () => {
      const wrapper = mount(Tag, {
        props: {
          closable: true
        }
      })
      const result = wrapper.findComponent(Tag).find('.el-tag__close').exists()

      expect(result).toBeTruthy()
    })
    it('disable-transitions', async () => {
      let disableTransitions = ref(false)
      const wrapper = mount(Tag, {
        props: {
          closable: true,
          disableTransitions
        }
      })

      expect(wrapper.find('transition-stub').exists()).toBeTruthy()
      await nextTick(() => {
        disableTransitions.value = true
      })
      expect(wrapper.find('transition-stub').exists()).toBeFalsy()
    })
    it('hit', () => {
      const wrapper = mount(Tag, {
        props: {
          hit: true
        }
      })
      const result = wrapper.findComponent(Tag).find('.is-hit').exists()

      expect(result).toBeTruthy()
    })
    it('color', () => {
      const wrapper = mount(Tag, {
        props: {
          color: 'red'
        }
      })
      const result = wrapper
        .findComponent(Tag)
        .find('[style="background-color: red;"]')
        .exists()

      expect(result).toBeTruthy()
    })
    it('size', () => {
      const size = 'small'
      const wrapper = mount(Tag, {
        props: {
          size
        }
      })
      const result = wrapper
        .findComponent(Tag)
        .find(`.el-tag--${size}`)
        .exists()

      expect(result).toBeTruthy()
    })
    it('size by global config', () => {
      const size = 'small'
      const wrapper = mount(Tag, {
        global: {
          plugins: [
            setupGlobalOptions({
              size
            })
          ]
        }
      })
      const result = wrapper
        .findComponent(Tag)
        .find(`.el-tag--${size}`)
        .exists()

      expect(result).toBeTruthy()
    })
    it('effect', () => {
      const effect = 'dark'
      const wrapper = mount(Tag, {
        props: {
          effect
        }
      })
      const result = wrapper
        .findComponent(Tag)
        .find(`.el-tag--${effect}`)
        .exists()

      expect(result).toBeTruthy()
    })
  })
  describe('events', () => {
    it('close', async () => {
      const wrapper = mount(Tag, {
        props: {
          closable: true
        }
      })
      await wrapper.find('.el-tag__close').trigger('click')

      expect(wrapper.emitted()).toHaveProperty('close')
      expect(wrapper.find('.el-tag').exists()).toBeFalsy()
    })
    it('click', async () => {
      const wrapper = mount(Tag)
      await wrapper.find('.el-tag').trigger('click')

      expect(wrapper.emitted()).toHaveProperty('click')
    })
  })
})
