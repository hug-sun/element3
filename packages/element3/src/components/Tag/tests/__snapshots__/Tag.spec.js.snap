// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag.vue snapshot 1`] = `
<transition-stub>
  <span
    class="el-tag el-tag--light"
  >
    <!---->
    <!---->
  </span>
</transition-stub>
`;
